### DISCUSSION TITLE
Activate hidden flags

### DISCUSSION THREAD

**Discussion by boernd** [prometheus-operator/prometheus-operator#4809] [Q&A] [answered]
With the next Prometheus release comes a new hidden flag which we want to use/test if it reduces CPU load when using many ServiceMonitors: https://github.com/prometheus/prometheus/pull/10634

I have not found a way to configure this in the Operator (just enableFeatures). Have I overlooked something or does this have to be implemented?

**Comment by simonpasquier** [comment]
Hmm the only option would be to use [strategic merge patch](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/strategic-merge-patch.md) to overwrite the list of args but you'll have to provide the full list...

**Reply by boernd** [reply]
I have found some issues regarding support of arbitrary command line flags (one of them: https://github.com/prometheus-operator/prometheus-operator/issues/412#issuecomment-308048657) but they have always been declined/closed. Does this still hold true today? Imo adding an extra args list var would not hurt and be straight forward to implement.

**Comment by simonpasquier** [comment, marked as answer]
Supporting additional args isn't so simple because the operator should then check that they don't conflict with the generated args. What Frederic said a while ago still holds true IMHO:

> The whole point of it is to abstract away, how you're running Prometheus. Specifically storage flags should be abstracted and are today. If there are other flags you want to set that are currently not possible but legitimately should be possible, then that's a missing feature

Now hidden flags are problematic: I'd be wary to support them since they have no stability guarantees.

**Reply by boernd** [reply]
So the operator will only support new flags if they are non-hidden ones? Or would you accept a PR for explicitly supporting this flag? Just fyi, setting this flag to 30s (default is 5s) decreased the Prometheus CPU usage in our setup from ~25 CPUs to about 8 (with ~700 Servicemonitors). So it is quite important for us to have this flag available in the near future.

**Reply by simonpasquier** [reply]
> So the operator will only support new flags if they are non-hidden ones?

Yes I'd be reluctant to add new fields to the CRD for hidden flags. Having said that, supporting all possible command line arguments in the CRD is also a challenge and we should probably have an easier way to customize the argument list than with [strategic merge patch](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/strategic-merge-patch.md). As I alluded before, container args can't be merged/appended with this technique so it's not really user-friendly.
Personally I think that we could have a `spec.additionalArgs` field where users would provide command-line arguments to be added to the generated list. Obviously you could still shoot yourself in the foot since the operator wouldn't do any sanity check wrt version support for instance but it's still less error-prone than with strategic merge patch IMHO.

**Reply by simonpasquier** [reply]
See #4851
